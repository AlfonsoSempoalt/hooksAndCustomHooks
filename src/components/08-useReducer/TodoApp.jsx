import React, { useEffect, useReducer } from 'react';
import { useForm } from '../../hooks/useForm';
import '../02-UseEffect/effects.css';
import todoReducer from './todoReducer';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);
  const [formValues, handleInputChange, resetForm] = useForm({
    description: '',
  });

  const { description } = formValues;

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim().length >= 2) {
      const newTodo = {
        id: new Date().getTime(),
        desc: description,
        done: false,
      };

      const action = {
        type: 'add',
        payload: newTodo,
      };

      dispatch(action);
      resetForm();
    }
  };

  const handleOnDelete = (todoId) => {
    const action = {
      type: 'delete',
      payload: todoId,
    };
    dispatch(action);
  };

  const handleToggle = (todoId) => {
    dispatch({
      type: 'toggle',
      payload: todoId,
    });
  };

  return (
    <>
      <h1>Todo App ({todos.length})</h1>
      <hr />
      <section className='row'>
        <section className='col-7'>
          <ol className='list-group list-group-flush'>
            {todos.map((todo, i) => (
              <li key={todo.id} className='list-group-item text-center'>
                <p
                  className={`${todo.done && 'complete'}`}
                  onClick={() => handleToggle(todo.id)}
                >
                  {i + 1} {todo.desc}
                </p>
                <button
                  className='btn btn-danger'
                  onClick={() => handleOnDelete(todo.id)}
                >
                  Delete
                </button>
              </li>
            ))}
          </ol>
        </section>
        <section className='col-5'>
          <h4>Add todo</h4>
          <hr />
          <form onSubmit={handleSubmit} className='d-grid gap-2'>
            <input
              type='text'
              placeholder='learn...'
              name='description'
              value={description}
              onChange={handleInputChange}
              autoComplete='off'
              className='form-control'
            />
            <button
              type='submit'
              className='btn btn-outline-primary mt-2 btn-block'
            >
              Add
            </button>
          </form>
        </section>
      </section>
    </>
  );
};

export default TodoApp;
