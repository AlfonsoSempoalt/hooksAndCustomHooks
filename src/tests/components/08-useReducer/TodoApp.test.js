import React from 'react';
import { shallow, mount } from 'enzyme';
import TodoApp from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../fixtures/demoTodos';
import { act } from '@testing-library/react';

describe('tests to <TodoApp />', () => {
    const wrapper = shallow(<TodoApp />);
    Storage.prototype.setItem = jest.fn(() => {});

    test('should match with the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should add a Todo', () => {
        const wrapper = mount(<TodoApp />);
        act(() => {
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
        });
        expect(wrapper.find('h1').text().trim()).toBe('Todo App (1)');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

    test('should delete a Todo', () => {
        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
        wrapper.find('TodoList').prop('handleOnDelete')(demoTodos[0].id);
        expect(wrapper.find('h1').text().trim()).toBe('Todo App (0)');
    });
});
