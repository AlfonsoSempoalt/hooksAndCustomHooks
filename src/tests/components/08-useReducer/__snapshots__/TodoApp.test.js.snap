// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests to <TodoApp /> should match with the snapshot 1`] = `
<Fragment>
  <h1>
    Todo App (
    0
    )
  </h1>
  <hr />
  <section
    className="row"
  >
    <section
      className="col-7"
    >
      <TodoList
        handleOnDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </section>
    <section
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </section>
  </section>
</Fragment>
`;
